src/github.com/matrix-org/dendrite/roomserver/input/membership.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package input
+
+import (
+	"fmt"
+
+	"github.com/matrix-org/dendrite/roomserver/api"
+	"github.com/matrix-org/dendrite/roomserver/types"
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+func updateMemberships(
+	db RoomEventDatabase, updater types.RoomRecentEventsUpdater, removed, added []types.StateEntry,
+) ([]api.OutputEvent, error) {
+	changes := membershipChanges(removed, added)
+	var eventNIDs []types.EventNID
+	for _, change := range changes {
+		if change.added.EventNID != 0 {
+			eventNIDs = append(eventNIDs, change.added.EventNID)
+		}
+		if change.removed.EventNID != 0 {
+			eventNIDs = append(eventNIDs, change.removed.EventNID)
+		}
+	}
+	events, err := db.Events(eventNIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	var updates []api.OutputEvent
+
+	for _, change := range changes {
+		var ae *gomatrixserverlib.Event
+		var re *gomatrixserverlib.Event
+		var targetNID types.EventStateKeyNID
+		if change.removed.EventNID != 0 {
+			ev, _ := eventMap(events).lookup(change.removed.EventNID)
+			if ev != nil {
+				re = &ev.Event
+			}
+			targetNID = change.removed.EventStateKeyNID
+		}
+		if change.added.EventNID != 0 {
+			ev, _ := eventMap(events).lookup(change.added.EventNID)
+			if ev != nil {
+				ae = &ev.Event
+			}
+			targetNID = change.added.EventStateKeyNID
+		}
+		if updates, err = updateMembership(updater, targetNID, re, ae, updates); err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+func updateMembership(
+	updater types.RoomRecentEventsUpdater, targetNID types.EventStateKeyNID,
+	remove *gomatrixserverlib.Event, add *gomatrixserverlib.Event,
+	updates []api.OutputEvent,
+) ([]api.OutputEvent, error) {
+	var err error
+	old := "leave"
+	new := "leave"
+
+	if remove != nil {
+		old, err = remove.Membership()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if add != nil {
+		new, err = add.Membership()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if old == new {
+		return updates, nil
+	}
+
+	mu, err := updater.MembershipUpdater(targetNID)
+	if err != nil {
+		return nil, err
+	}
+
+	return updateWithNewMembership(mu, new, add, updates)
```

Comment:
You call this function in this place only and you just made a `MembershipUpdater` in order to pass it to this function. Why don't you put `updateWithNewMembership` as a function on `MembershipUpdater`?